src/geometry/partitions/distributedptscotchunstructuredpartition.h

```diff
@@ -0,0 +1,288 @@
+            const Coord<2>& resolution = Coord<2>(50, 50))
+{
+    Coord<2> from = region.boundingBox().origin;
+    Coord<2> dim = region.boundingBox().dimensions;
+    for (int y = 0; y < resolution.y(); ++y) {
+        for (int x = 0; x < resolution.x(); ++x) {
+            int cx = from.x() + ((float)x / (float)resolution.x()) * dim.x();
+            int cy = from.x() + ((float)y / (float)resolution.y()) * dim.y();
+
+            if (region.count(Coord<2>(cx, cy)) > 0) {
+                os << '#';
+            } else {
+                os << '.';
+            }
+        }
+        os << '\n';
+    }
+}
+
+template<int DIM>
+class DistributedPTScotchUnstructuredPartition : public Partition<DIM>
+{
+public:
+    using Partition<DIM>::startOffsets;
+    using Partition<DIM>::weights;
+
+    DistributedPTScotchUnstructuredPartition(
+            const Coord<DIM>& origin,
+            const Coord<DIM>& dimensions,
+            const long offset,
+            const std::vector<std::size_t>& weights,
+            const Adjacency& adjacency) :
+        Partition<DIM>(offset, weights),
+        origin(origin),
+        dimensions(dimensions),
+        numPartitions(weights.size())
+    {
+        this->adjacency = adjacency;
+
+        getStartEnd(MPILayer().rank(), &start, &end);
+        localCells = end - start;
+        buildRegions();
+    }
+
+    DistributedPTScotchUnstructuredPartition(
+            const Coord<DIM>& origin,
+            const Coord<DIM>& dimensions,
+            const long offset,
+            const std::vector<std::size_t>& weights,
+            Adjacency&& adjacency) :
+        Partition<DIM>(offset, weights),
+        origin(origin),
+        dimensions(dimensions),
+        numPartitions(weights.size())
+    {
+        this->adjacency = std::move(adjacency);
+
+        getStartEnd(MPILayer().rank(), &start, &end);
+        localCells = end - start;
+        buildRegions();
+    }
+
+    void getStartEnd(int rank, std::size_t *start, std::size_t *end)
+    {
+        std::size_t cells = dimensions.prod();
+        *start = (float)rank / numPartitions * cells;
+        *end = (float)(rank + 1) / numPartitions * cells;
+    }
+
+    Region<DIM> getRegion(const std::size_t node) const override
+    {
+        return regions.at(node);
+    }
+
+private:
+    Coord<DIM> origin;
+    Coord<DIM> dimensions;
+    SCOTCH_Num localCells = 0;
+    std::size_t start = 0, end = 0;
+    std::size_t numPartitions = 0;
+    std::vector<Region<DIM> > regions;
+
+    void buildRegions()
+    {
+        std::vector<SCOTCH_Num> indices(localCells);
+        initIndices(indices);
+
+        regions.resize(numPartitions);
+        createRegions(indices);
+    }
+
+    void initIndices(std::vector<SCOTCH_Num>& indices)
+    {
+        // create 2D grid
+        SCOTCH_Dgraph graph;
+        int error = SCOTCH_dgraphInit(&graph, MPILayer().communicator());
+
+        SCOTCH_Num numEdges = 0;
+
+        // ok, this is SUPER ugly
```

Comment:
maybe RegionBasedAdjacency should have  a member size() that refers to Region::size()?